cranelift/codegen/src/souper_harvest.rs

```diff
@@ -0,0 +1,525 @@
+//! Harvest left-hand side superoptimization candidates.
+//!
+//! Given a clif function, harvest all its integer subexpressions, so that they
+//! can be fed into [Souper](https://github.com/google/souper) as candidates for
+//! superoptimization. For some of these candidates, Souper will successfully
+//! synthesize a right-hand side that is equivalent but has lower cost than the
+//! left-hand side. Then, we can combine these left- and right-hand sides into a
+//! complete optimization, and add it to our peephole passes.
+//!
+//! To harvest the expression that produced a given value `x`, we do a
+//! post-order traversal of the dataflow graph starting from `x`. As we do this
+//! traversal, we maintain a map from clif values to their translated Souper
+//! values. We stop traversing when we reach anything that can't be translated
+//! into Souper IR: a memory load, a float-to-int conversion, a block parameter,
+//! etc. For values produced by these instructions, we create a Souper `var`,
+//! which is an input variable to the optimization. For instructions that have a
+//! direct mapping into Souper IR, we get the Souper version of each of its
+//! operands and then create the Souper version of the instruction itself. It
+//! should now be clear why we do a post-order traversal: we need an
+//! instruction's translated operands in order to translate the instruction
+//! itself. Once this instruction is translated, we update the clif-to-souper
+//! map with this new translation so that any other instruction that uses this
+//! result as an operand has access to the translated value. When the traversal
+//! is complete we return the translation of `x` as the root of left-hand side
+//! candidate.
+
+use crate::ir;
+use souper_ir::ast;
+use std::collections::{HashMap, HashSet};
+use std::string::String;
+use std::sync::mpsc;
+use std::vec::Vec;
+
+/// Harvest Souper left-hand side candidates from the given function.
+///
+/// Candidates are reported through the given MPSC sender.
+pub fn do_souper_harvest(func: &ir::Function, out: &mut mpsc::Sender<String>) {
+    let mut allocs = Allocs::default();
+
+    // Iterate over each instruction in each block and try and harvest a
+    // left-hand side from its result.
+    for block in func.layout.blocks() {
+        let mut option_inst = func.layout.first_inst(block);
+        while let Some(inst) = option_inst {
+            let results = func.dfg.inst_results(inst);
+            if results.len() == 1 {
```

Comment:
I don't remember how many clif instructions have >1 result, but in the future this could be loosened potentially be loosened up to looking at all results and possibly materializing the results from different Souper-found optimizations, right?

Reply:
There aren't a ton that have multiple results. The most common are multi-value calls, which aren't relevant here. But add-with-carry and combined div/mul are candidates.

We would also need to extend peepmatic to work with operations that return multiple values, which it doesn't support right now. I expect this would require more effort than the harvester, which is relatively small and isolated.